# CSC495 - DataMining - HW#1 - Rinty Chowdhury

In [122]:
import pandas as pd
import numpy as np
import scipy
import sklearn
from scipy import stats
from sklearn.preprocessing import MinMaxScaler

In [123]:
my_data = pd.read_csv('../AssignmentMaterials/Video_Store.tsv', sep='\t', header=0)
my_data.head()

Cust ID Gender   Income  Age  Rentals  Avg Per Visit Incidentals   Genre
0        1      M  $45,000   25       27            2.5         Yes  Action
1        2      F  $54,000   33       12            3.4          No   Drama
2        3      F  $32,000   20       42            1.6          No  Comedy
3        4      F  $59,000   70       16            4.2         Yes   Drama
4        5      M  $37,000   35       25            3.2         Yes  Action

## 1. There are seven features in the data set.

## 2. There are four continuous features, and three nominal features.

## 3. Four continuous features are income, age, rentals, and avg per visit. Calculation of mean, median, maximum, minimum, and SD are below.

In [124]:
# Removing "$", "," from Income column and coverting it from string to float
my_data["Income"] = my_data["Income"].str.replace('$','')
my_data["Income"] = my_data["Income"].str.replace(',','')
my_data['Income'] = my_data['Income'].astype(float)

In [125]:
# Calculation for Income 
print("Mean income: " + str(my_data["Income"].mean()))
print("Median income: " + str(my_data["Income"].median()))
print("Maximum income: " + str(my_data["Income"].max()))
print("Minimum income: " + str(my_data["Income"].min()))
print("STD income: " + str(my_data["Income"].std()))
print()

# Calculation for Age 
print("Mean age: " + str(my_data["Age"].mean()))
print("Median age: " + str(my_data["Age"].median()))
print("Maximum age: " + str(my_data["Age"].max()))
print("Minimum age: " + str(my_data["Age"].min()))
print("STD age: " + str(my_data["Age"].std()))
print()

# Calculation for Rentals
print("Mean rentals: " + str(my_data["Rentals"].mean()))
print("Median rentals: " + str(my_data["Rentals"].median()))
print("Maximum rentals: " + str(my_data["Rentals"].max()))
print("Minimum rentals: " + str(my_data["Rentals"].min()))
print("STD rentals: " + str(my_data["Rentals"].std()))
print()

# Calculation for Avg_Per_Visit
print("Mean avg per visit: " + str(my_data["Avg Per Visit"].mean()))
print("Median avg per visit: " + str(my_data["Avg Per Visit"].median()))
print("Maximum avg per visit: " + str(my_data["Avg Per Visit"].max()))
print("Minimum avg per visit: " + str(my_data["Avg Per Visit"].min()))
print("STD avg per visit: " + str(my_data["Avg Per Visit"].std()))

Mean income: 42300.0
Median income: 41000.0
Maximum income: 89000.0
Minimum income: 1000.0
STD income: 21409.753642216627

Mean age: 31.56
Median age: 30.0
Maximum age: 70
Minimum age: 15
STD age: 12.000272105758475

Mean rentals: 26.24
Median rentals: 25.0
Maximum rentals: 48
Minimum rentals: 9
STD rentals: 10.027635283941828

Mean avg per visit: 2.748
Median avg per visit: 2.75
Maximum avg per visit: 4.7
Minimum avg per visit: 1.1
STD avg per visit: 0.8981250310667058


## 4.1 Smoothing by bin means with bin depth of 4 of the 'Age' attribute

In [126]:
# Sort the age
sorted_age = my_data["Age"].values
# Create numpy array using the sorted age
sorted_age = np.sort(sorted_age)
# Calculate the bin means, bin intervals, and bin labels using scipy.stats
mean_ages, bin_intervals, bin_labels = scipy.stats.binned_statistic(my_data.index, sorted_age, statistic='mean', bins=12)
# Remove 0 from the index and add the maximum value of the Age column into the intervals
bin_intervals = np.delete(bin_intervals, 0)
bin_intervals = np.append(bin_intervals, my_data["Age"].max())
# Create the new column with the calculated mean
my_data['Mean Age'] = pd.cut(x=my_data['Age'], bins=bin_intervals, labels=mean_ages)

## 4.2 Min-max normalization of the 'Income' attribute

In [127]:
# Scale the data using MinMaxcScaler
scaler = MinMaxScaler()
# Reshape the data column in order to fit the scaler
data = my_data["Income"].values.reshape(-1,1)
# Fit the data into the scaler and scale the value
scaler.fit(data)
# Create new column with the scaled value
my_data['Normalized Income'] = scaler.transform(data)

## 4.3 Z-score normalization of the 'Rentals' attribute

In [128]:
# Normalizing the rentals value using z-score and creating new column with that value
my_data['Normalized Rentals'] = scipy.stats.mstats.zscore(my_data["Rentals"])

## 4.4 Discretize the original 'Income' attribute

In [129]:
# Create new column based on the given bin intervals and labels
my_data["Income Category"] = pd.cut(x=my_data['Income'], bins=[0, 25000, 59000, my_data["Income"].max()], labels=["Low", "Mid", "High"])
my_data.head(50)

Cust ID Gender   Income  Age  Rentals  Avg Per Visit Incidentals   Genre  \
0         1      M  45000.0   25       27            2.5         Yes  Action   
1         2      F  54000.0   33       12            3.4          No   Drama   
2         3      F  32000.0   20       42            1.6          No  Comedy   
3         4      F  59000.0   70       16            4.2         Yes   Drama   
4         5      M  37000.0   35       25            3.2         Yes  Action   
5         6      M  18000.0   20       33            1.7          No  Action   
6         7      F  29000.0   45       19            3.8          No   Drama   
7         8      M  74000.0   25       31            2.4         Yes  Action   
8         9      M  38000.0   21       18            2.1          No  Comedy   
9        10      F  65000.0   40       21            3.3          No   Drama   
10       11      F  41000.0   22       48            2.3         Yes   Drama   
11       12      F  26000.0   22       29            2.9         Yes  Action   
12       13      M  83000.0   46       14            3.6          No  Comedy   
13       14      M  45000.0   36       24            2.7          No   Drama   
14       15      M  68000.0   30       36            2.7         Yes  Comedy   
15       16      M  17000.0   19       26            2.2         Yes  Action   
16       17      M  36000.0   35       28            3.5         Yes   Drama   
17       18      F   6000.0   16       39            1.8         Yes  Action   
18       19      F  24000.0   25       41            3.1          No  Comedy   
19       20      M  12000.0   16       23            2.2         Yes  Action   
20       21      F  47000.0   52       11            3.1          No   Drama   
21       22      M  25000.0   33       16            2.9         Yes   Drama   
22       23      F   2000.0   15       30            2.5          No  Comedy   
23       24      F  79000.0   35       22            3.8         Yes   Drama   
24       25      M   1000.0   16       25            1.4         Yes  Comedy   
25       26      F  56000.0   35       40            2.6         Yes  Action   
26       27      F  62000.0   47       32            3.6          No   Drama   
27       28      M  57000.0   52       22            4.1          No  Comedy   
28       29      F  15000.0   18       37            2.1         Yes  Action   
29       30      M  41000.0   25       17            1.4         Yes  Action   
30       31      F  49000.0   56       15            3.2          No  Comedy   
31       32      M  47000.0   30       21            3.1         Yes   Drama   
32       33      M  23000.0   25       28            2.7          No  Action   
33       34      F  29000.0   32       19            2.9         Yes  Action   
34       35      M  74000.0   29       43            4.6         Yes  Action   
35       36      F  29000.0   21       34            2.3          No  Comedy   
36       37      M  89000.0   46       12            1.2          No  Comedy   
37       38      M  41000.0   38       20            3.3         Yes   Drama   
38       39      F  68000.0   35       19            3.9          No  Comedy   
39       40      M  17000.0   19       32            1.8          No  Action   
40       41      F  50000.0   33       17            1.4          No   Drama   
41       42      M  32000.0   25       26            2.2         Yes  Action   
42       43      F  49000.0   28       48            3.3         Yes   Drama   
43       44      M  35000.0   24       24            1.7          No   Drama   
44       45      M  56000.0   38       30            3.5         Yes   Drama   
45       46      F  57000.0   43        9            1.1          No   Drama   
46       47      F  69000.0   35       22            2.8         Yes   Drama   
47       48      F  52000.0   47       14            1.6          No   Drama   
48       49      M  31000.0   25       42            3.4         Yes  Action   
49  

## 5. Converting categorical attribute into multiple attributes


In [130]:
my_data = pd.read_csv('./AssignmentMaterials/Video_Store.tsv', sep='\t', header=0)
my_data["Income"] = my_data["Income"].str.replace('$','')
my_data["Income"] = my_data["Income"].str.replace(',','')
my_data['Income'] = my_data['Income'].astype(float)
# Creating new column based on the given column names and prefix
my_data = pd.get_dummies(my_data, columns=["Gender", "Incidentals", "Genre"], prefix=["Gender", "Incidentals", "Genre"])
my_data.head()

Cust ID   Income  Age  Rentals  Avg Per Visit  Gender_F  Gender_M  \
0        1  45000.0   25       27            2.5         0         1   
1        2  54000.0   33       12            3.4         1         0   
2        3  32000.0   20       42            1.6         1         0   
3        4  59000.0   70       16            4.2         1         0   
4        5  37000.0   35       25            3.2         0         1   

   Incidentals_No  Incidentals_Yes  Genre_Action  Genre_Comedy  Genre_Drama  
0               0                1             1             0            0  
1               1                0             0             0            1  
2               1                0             0             1            0  
3               0                1             0             0            1  
4               0                1             1             0            0

## 6. Correlation matrix and analysis

In [131]:
# Deleting 'Cust ID' column
del(my_data['Cust ID'])
# Creating correlation matrix
correlation_matrix = my_data.corr()
correlation_matrix

Income       Age   Rentals  Avg Per Visit  Gender_F  \
Income           1.000000  0.627013 -0.274873       0.305221  0.063841   
Age              0.627013  1.000000 -0.549611       0.397372  0.227663   
Rentals         -0.274873 -0.549611  1.000000       0.070302  0.025164   
Avg Per Visit    0.305221  0.397372  0.070302       1.000000  0.029176   
Gender_F         0.063841  0.227663  0.025164       0.029176  1.000000   
Gender_M        -0.063841 -0.227663 -0.025164      -0.029176 -1.000000   
Incidentals_No   0.063841  0.214184 -0.297452      -0.078885  0.198718   
Incidentals_Yes -0.063841 -0.214184  0.297452       0.078885 -0.198718   
Genre_Action    -0.362514 -0.480794  0.326052      -0.101414 -0.303576   
Genre_Comedy     0.071588  0.016872 -0.032456      -0.067208  0.022496   
Genre_Drama      0.292781  0.456371 -0.291170       0.157955  0.277831   

                 Gender_M  Incidentals_No  Incidentals_Yes  Genre_Action  \
Income          -0.063841        0.063841        -0.063841     -0.362514   
Age             -0.227663        0.214184        -0.214184     -0.480794   
Rentals         -0.025164       -0.297452         0.297452      0.326052   
Avg Per Visit   -0.029176       -0.078885         0.078885     -0.101414   
Gender_F        -1.000000        0.198718        -0.198718     -0.303576   
Gender_M         1.000000       -0.198718         0.198718      0.303576   
Incidentals_No  -0.198718        1.000000        -1.000000     -0.386976   
Incidentals_Yes  0.198718       -1.000000         1.000000      0.386976   
Genre_Action     0.303576       -0.386976         0.386976      1.000000   
Genre_Comedy    -0.022496        0.397430        -0.397430     -0.421464   
Genre_Drama     -0.277831        0.032686        -0.032686     -0.612372   

                 Genre_Comedy  Genre_Drama  
Income               0.071588     0.292781  
Age                  0.016872     0.456371  
Rentals             -0.032456    -0.291170  
Avg Per Visit       -0.067208     0.157955  
Gender_F             0.022496     0.277831  
Gender_M            -0.022496    -0.277831  
Incidentals_No       0.397430     0.032686  
Incidentals_Yes     -0.397430    -0.032686  
Genre_Action        -0.421464    -0.612372  
Genre_Comedy         1.000000    -0.458831  
Genre_Drama         -0.458831     1.000000

### Results:
Based on the above correlation matrix, male gender and female gender columns has very strong negative correlation which is expected. Because one customer can only be either male or female. The correlation value is -1. Also, Incidentals_No and Incidentals_Yes has very strong negative correlation which is also expected. Because one customer will either buy incidentals or not. The correlation value is -1. Those are the only two strongly correlated features based on the calculated correlation matrix. Other than that, there are some other attributes that are significantly correlated with each other. Income and Age has moderate positive correlation. When Age goes high, Income will also increase significantly. The correlation value is 0.627013. Additionly, Genre_Drama and Genre_Action has moderate negative correlation. When customers rent more drama video then action video rentals get lower significantly. The correlation value is -0.612372.